cypress-io/cypress
# fix: Preserve tree expansion state

Video: https://cypressio.slack.com/archives/C035VB3HVAL/p1655219943591069

Addressing a reactivity issue we currently have in our `useCollapsibleTree` usage. In ACI, row data is loaded as the user scrolls and periodically in the background - these fetches flow back through GQL and present as a "new" list of specs (even though it's the same specs, just with new deeply-nested content) which triggers the tree data to get rebuilt. This manifests as the tree state re-setting on scroll and is likely contributing to some of the scroll lagginess we're seeing.

Without reworking the implementation which would impact multiple UI areas at present this is a next-best: incorporate an external cache that can be used to persist the expansion state of the directories so that when the tree does get rebuilt the state is maintained. This cache is then cleared when the user searches or changes the list of specs because we want the directories to expand in those situations to match 10.1.0 behavior.

### User facing changelog


### Additional details

I wasn't able to find a good way to test this in CT; since SpecList is using `mountFragment` and GQL I couldn't find a way to mutate the GQL data and trigger a reactive data update within a single mount cycle. If someone knows how to do this I'd love to learn how

### Steps to test

1. Open E2E mode with a project that has at least one recorded run in the cloud (preferably one with 
2. Log in if not already
3. Observe latest run data begin to load
4. Collapse a directory and scroll down
5. Observe directory remains collapsed as lower rows scroll into viewport and load cloud data
	- Current behavior would result in directory resetting to expanded

### How has the user experience changed?
SpecsList directory expansion state should persist as ACI data is loaded during scroll

### PR Tasks

- [na] Has the original issue (or this PR, if no issue exists) been tagged with a release in ZenHub? (user-facing changes only)
- [na] Has a PR for user-facing changes been opened in [`cypress-documentation`](https://github.com/cypress-io/cypress-documentation)? 
- [na] Have API changes been updated in the [`type definitions`](https://github.com/cypress-io/cypress/blob/develop/cli/types/cypress.d.ts)?
